### DISCUSSION TITLE
Redirecting based on a condition

### DISCUSSION THREAD

**Discussion by jongallant** [dotnet/yarp#1937] [Q&A] [answered]
Hey, just curious if I am missing something obvious.  I am attempting run custom validation in a middleware, and then if certain conditions are met to redirect to a different destination than requested.  

Essentially what I am trying to do is perform a response redirect, so that the user is directed to a different location.

What is the way to do this correctly with YARP?  Am I forced to use Direct Forwarding?

**Comment by MihaZupan** [comment, marked as answer]
There's a `context.ReassignProxyRequest(cluster)` API you can use to change the destination in a middleware.
See https://microsoft.github.io/reverse-proxy/articles/ab-testing.html

**Reply by jongallant** [reply]
Hey thanks.  I am not trying to send it to a different cluster, just a different resource on the same cluster.

**Reply by MihaZupan** [reply]
You have a few options.
- Since you're in a middleware already, you could just short-circuit the response and return a redirect, forcing the client to fetch it again.
- If you just want to modify just the path, you can alter the `httpContext.Request` and let YARP proxy to that resource instead.
- The cleanest option may be to just do this in a [request transform](https://microsoft.github.io/reverse-proxy/articles/transforms.html#from-code)
  ```c#
  services.AddReverseProxy()
      .AddTransforms(transformBuilderContext =>
      {
          transformBuilderContext.AddRequestTransform(async context =>
          {
              if (await MyConditionAsync(context))
              {
                  var newUri = new UriBuilder(context.ProxyRequest.RequestUri);
                  newUri.Path = "/something";
                  context.ProxyRequest.RequestUri = newUri.Uri;
              }
          });
      })
  ```
